### PR TITLE
MINOR: distinguish between missing source topics and internal assignment errors

### PR DESCRIPTION
**Pull request by ableegoldman**
Minor followup to KAFKA-10559

I noticed that we were converting any and all TaskAssignmentException to the INCOMPLETE_SOURCE_TOPIC_METADATA error code to shut down all the clients. Since these errors are typically fatal, it does seem appropriate to propagate the shutdown command. But we should do so with a new AssignorError instead of piggy-backing on the INCOMPLETE_SOURCE_TOPIC_METADATA, which would be pretty confusing for users who do in fact have all their source topics.

Changes in this PR:
- Add new AssignorError.ASSIGNMENT_ERROR for generic assignment errors
- Missing source topics --> throw/catch MissingSourceTopicException --> INCOMPLETE_SOURCE_TOPIC_METADATA 
- Internal assignment errors --> throw/catch TaskAssignmentException --> ASSIGNMENT_ERROR

Should be cherry-picked to 2.7

### CODE REVIEW

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java:346]: The diff is kind of annoying but basically I just moved all the try-catch blocks into a single outer try that encapsulates all of the assignment logic. If we throw a TaskAssignmentException at any point it'll bail and just encode the `ASSIGNMENT_ERROR` code (or `INCOMPLETE_SOURCE_TOPIC_METADATA` if MissingSourceTopicException is thrown)

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java:445]: Just tried to standardize all permanent errors to throw TaskAssignmentException instead. eg if there are incompatible mixed versions we should shut everyone down until the operator can sort it out

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java:716]: If we ever hit this it means there is a serious bug in the assignment algorithm. In that case we should go ahead and shut everyone down once we notice, since otherwise it would be a long slow death of single threads at a time

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java:1247]: Likewise, this means a bug in the assignment algorithm that should never be hit

**Comment by ableegoldman** [comment]
heads up @mjsax @guozhangwang

**Review by mjsax — APPROVED** [review]
Thanks @ableegoldman. LGTM, assuming Jenkins passes.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsRebalanceListener.java
Comment (vvcephei) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsRebalanceListener.java:66]: Hey @ableegoldman , I just looked into this, and it appears that this is the only place that the error code would be read and thrown as an exception to kill the thread, right?

If so, then it looks like older clients are looking specifically for the error code to be `== AssignorError.INCOMPLETE_SOURCE_TOPIC_METADATA.code()`, and they'd interpret any other code as "looks good" and proceed with PARTITIONS_ASSIGNED.

Maybe there's nothing we can do about it now, but perhaps we should put in a block to future-proof this code by adding an `else if (assignmentErrorCode.get() != 0) { throw new TaskAssignmentException("Unknown error code: "+assignmentErrorCode.get()) }` ?

Or is this already handled in some other way I'm not seeing?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java
Comment (vvcephei) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java:420]: Following on my other comment, should we gate this on the usedMetadataVersion so that we'd return a `INCOMPLETE_SOURCE_TOPIC_METADATA` for versions 7 and below, to ensure they'll be properly handled as exceptions on the member side?

**Review by vvcephei — COMMENTED** [review]
Thanks, @ableegoldman ! This LGTM; just one concern.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java:420]: I'm honestly not sure that INCOMPLETE_SOURCE_TOPIC_METADATA is better than just throwing nothing for older clients. Is it better to shut everyone down immediately if it not only completely obfuscates the reason for shutting down, but actually points to a completely different (and wrong) root cause?

Obviously a slow death by one thread at a time is not a good user experience, but this is slightly different. For one thing, you'd still see all the upgraded clients shut down, and only the older ones would remain. Presumably you're in the middle of a rolling bounce, so eventually all of those surviving older clients will eventually be upgraded to understand this error code and shut down. Note that the previous behavior is to just kill a single thread at a time upon hitting this exception, so it's not a regression. Throwing a misleading exception actually seems more like a regression to me. 

WDYT? How important does it seem to ensure that the entire group shuts down immediately during a rolling bounce? It does seem preferable to cut the rolling bounce early if it's doomed from the start. But I personally feel like logging a misleading error is a worse user experience.

Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java:420]: In my experience, users would not check the logs of every single instance, especially if the first set of logs they checked states very clearly what the problem is -- even if it makes no sense

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsRebalanceListener.java
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsRebalanceListener.java:66]: +1 on future proofing this

Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsRebalanceListener.java:66]: Ooh, but we did use to have a VERSION_PROBING error code that had value `2`. So we should skip that and go right to `3`, thanks for reminding me

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java:417]: I could not find a unit test that verifies this code.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java:421]: I could not find a unit test that verifies this code.

**Review by cadonna — COMMENTED** [review]
@ableegoldman Thanks for the PR.

I could not find unit tests for some code.

Additionally, I propose that we should ensure that the relevant system tests are green before somebody merges this PR.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java
Comment (vvcephei) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java:420]: Perhaps I did misunderstand the code after all.

My read was that the old clients would actually _not_ get an exception if we return a new code. Instead, they would continue running in an undefined state and maybe just cryptically do nothing, or maybe throw an even more confusing exception, or maybe do something worse.

That was my motivation for making sure they would at least throw an exception (the same exception they've always thrown for this condition, right?)

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java:417]: Ah. I had assumed there was one because this functionality is so old, but nope. Thanks for actually checking (added one)

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java:421]: This is covered by `shouldEncodeAssignmentErrorIfPreVersionProbingSubscriptionAndFutureSubscriptionIsMixed` which hits the TaskAssignmentException in `checkMetadataVersions()`.

I can try to add another test for one of the other instances where TaskAssignmentException might be thrown, but I would have to go really out of the way to artificially corrupt the data structures of the assignor in order to actually hit it (eg by mocking the `Cluster` so that it returns one set of topics the first time it's called, and then reports another set of topics the next time). I'm honestly not sure if it adds any value at that point -- WDYT?

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignorTest.java
Comment (ableegoldman) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignorTest.java:1941]: I happened to notice that this test was not actually using this `streamsBuilder`, so I tried to fix that and it broke. After some debugging I realized this test was broken in so many different ways that it's a miracle it was ever passing at all. Luckily all the bugs were in the test and not in the actual code 🙏

Comment (vvcephei) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignorTest.java:1941]: Goodness... Thanks for fixing this.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java
Comment (vvcephei) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java:420]: Ok, what I didn't realize before is that the "prior" code in this diff hasn't been released yet, so we are good to just do this change in place.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsRebalanceListener.java
Comment (vvcephei) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsRebalanceListener.java]: Maybe we shouldn't say "error" when it's not an error. I'm just imagining the mailing list questions that will start pouring in...

```suggestion
            log.info("Received version probing code {}", AssignorError.VERSION_PROBING);
```

**Review by vvcephei — APPROVED** [review]
LGTM! Thanks, @ableegoldman . 

I just had one minor thought about the log message. Thanks for this PR.

**Comment by ableegoldman** [comment]
Merged to trunk

**Comment by ableegoldman** [comment]
Cherry-picked to 2.7 (had to fix up a test in StreamsPartitionAssignorTest that was relying on the new `ReferenceContainer` which is only in trunk)

**Pull request merged**
